reference/src/clientimpl/src/sa_key_provision.c

```diff
@@ -16,349 +16,205 @@
+             return SA_STATUS_NULL_PARAMETER;
          }
          break;
-      default:
-         break;
-  }
-
 
-  return status;
+      case APPLE_MFI_PROVISIONING:
+         if (NULL == ((AppleMfiProvisioning*)in)->mfiBaseKey            ||
+             0 == ((AppleMfiProvisioning*)in)->mfiBaseKeyLength         ||
+             NULL == ((AppleMfiProvisioning*)in)->mfiProvisioningObject ||
+             0 == ((AppleMfiProvisioning*)in)->mfiProvisioningObjectLength) {
+             ERROR("null private key/object or size is  0");
+             return SA_STATUS_NULL_PARAMETER;
+         }
+         /*TODO SoC Vendor:here just provide an example to pass-through MFI base key
+           and provisioning obejet,
+           how to use these data is all up to SOC vendor.
+          */
+        {
+           void   *in_mfiBaseKey =
+                  ((AppleMfiProvisioning*)in)->mfiBaseKey;
+           size_t in_mfiBaseKey_length =
+                  ((AppleMfiProvisioning*)in)->mfiBaseKeyLength;
+           void   *in_provisioningObject =
+                  ((AppleMfiProvisioning*)in)->mfiProvisioningObject;
+           size_t in_provisioningObject_length =
+                  ((AppleMfiProvisioning*)in)->mfiProvisioningObjectLength;
+           INFO("in_mfiBaseKey:0x%x, in_mfiBaseKey_length:%d,\
+                  in_provisioningObject:0x%x, in_provisioningObject_length:%d",
+                  in_mfiBaseKey, in_mfiBaseKey_length,
+                  in_provisioningObject, in_provisioningObject_length);
+        }
+        return status;
+
+     case APPLE_FAIRPLAY_PROVISIONING:
+        if (NULL == ((AppleFairPlayProvisioning*)in)->fairPlaySecret  ||
+            0 == ((AppleFairPlayProvisioning*)in)->fairPlaySecretLength) {
+            ERROR("null private key or size is  0");
+            return SA_STATUS_NULL_PARAMETER;
+        }
+        /*TODO SoC Vendor:here just provide an example to pass-through fairplay secret key
+          how to use these data is all up to SOC vendor.
+         */
+        {
+            void   *in_fairPlaySecret =
+                   ((AppleFairPlayProvisioning*)in)->fairPlaySecret;
+            size_t in_fairPlaySecret_length =
+                   ((AppleFairPlayProvisioning*)in)->fairPlaySecretLength;
+            INFO("in_fairPlaySecret:0x%x, in_fairPlaySecret_length:%d",
+                   in_fairPlaySecret, in_fairPlaySecret_length);
+        }
+        return status;
+
+     default:
+        ERROR("unknown DRM provisioning type");
+        return SA_STATUS_INVALID_PARAMETER;
+    }
+ 
+    return status;
 }
 
+sa_status sa_key_provision_ta (
+   sa_key_type_ta ta_key_type,
+   const void* in,
+   size_t in_length,
+   void* parameters) {
+
+   sa_status status = SA_STATUS_OK;
+   if (SA_STATUS_OK != 
+      (status = sanityCheck(ta_key_type, in, in_length, parameters))) {
+      return status; 
+   }
+
+   /*TODO SoC Vendor: here is the place to call your TA to provision the key
```

Comment:
"TODO SoC Vendor" code all needs to be in an "src/internal" or "src/porting" file.   Code that is not located in internal or porting should not be changed and is considered platform independent.